serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt

```diff
@@ -2,154 +2,55 @@ package info.mqtt.android.service.ping
 
 import org.eclipse.paho.client.mqttv3.MqttPingSender
 import org.eclipse.paho.client.mqttv3.internal.ClientComms
-import android.content.BroadcastReceiver
-import android.app.PendingIntent
-import kotlin.jvm.Volatile
 import timber.log.Timber
-import android.content.IntentFilter
-import android.content.Intent
-import android.app.AlarmManager
-import java.lang.IllegalArgumentException
 import android.os.SystemClock
-import android.os.Build
-import org.eclipse.paho.client.mqttv3.IMqttToken
-import org.eclipse.paho.client.mqttv3.IMqttActionListener
-import org.eclipse.paho.client.mqttv3.MqttException
-import java.lang.Exception
-import android.annotation.SuppressLint
-import android.app.Service
-import android.content.Context
-import android.os.PowerManager
+import androidx.work.*
 import info.mqtt.android.service.MqttService
-import info.mqtt.android.service.MqttServiceConstants
-import kotlinx.coroutines.*
-import kotlin.system.measureTimeMillis
+
+import java.util.concurrent.TimeUnit
+
 
 /**
  * Default ping sender implementation on Android. It is based on AlarmManager.
  *
- * This class implements the [MqttPingSender] pinger interface
+ * This class implements the [MqttPingSender] ping interface
  * allowing applications to send ping packet to server every keep alive interval.
  *
  * @see MqttPingSender
  */
 internal class AlarmPingSender(val service: MqttService) : MqttPingSender {
+    private var continuation: Operation? = null
     private var clientComms: ClientComms? = null
-    private var alarmReceiver: BroadcastReceiver? = null
-    private var pendingIntent: PendingIntent? = null
-
-    @Volatile
-    private var hasStarted = false
+    private val workManager = WorkManager.getInstance(service)
 
     override fun init(comms: ClientComms) {
         this.clientComms = comms
-        alarmReceiver = AlarmReceiver()
     }
 
     override fun start() {
-        val action = MqttServiceConstants.PING_SENDER + clientComms!!.client.clientId
-        Timber.d("Register AlarmReceiver to MqttService$action")
-        service.registerReceiver(alarmReceiver, IntentFilter(action))
-        pendingIntent = PendingIntent.getBroadcast(service, 0, Intent(action), PendingIntent.FLAG_UPDATE_CURRENT)
-        schedule(clientComms!!.keepAlive)
-        hasStarted = true
+        val pingRepeatWorkRequest = PeriodicWorkRequest
+            .Builder(PingWorker::class.java, clientComms!!.keepAlive / 100, TimeUnit.MILLISECONDS)
+            .addTag(PING_WORKER_TAG)
```

Comment:
You add a tag just to be able to cancel the work, but except that it is unique, so you can cancel it with the unique work name